### PR TITLE
upgrading artifacts to deblur 1.0.4

### PR DESCRIPTION
**Pull request by antgonza**
I'll test in qiitarc but want to have a review before hand. Thanks!

### CODE REVIEW

File: support_files/patches/180214_fix_upper_case.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    with TRN:
+        ftps = {ft: (fid, fp) for fid, fp, ft in a.filepaths
+                if ft in ['biom', 'preprocessed_fasta']}
+        biom = ftps['biom'][1]
+        t = load_table(biom)
+        current = t.ids('observation')
+        updated = map(lambda x: x.upper(), current)
+        if set(current) ^ set(updated):
+            print '*********>\nChanging biom: ', a.id, fp
+                t.update_ids({i: i.upper() for i in t.ids('observation')},
```
Comment (sjanssen2) [inline support_files/patches/180214_fix_upper_case.py]: I am not familiar enough with biom, but I think you can encounter a situation where you have two observations like 'ACGt' and 'ACGT' that were two separate rows in the old table, but now need to be collapsed, i.e. their read counts for the samples need to be summed, to the single observation 'ACGT'.
Does you script account for this necessary collapsing of counts?

Comment (antgonza) [reply support_files/patches/180214_fix_upper_case.py]: The code doesn't account for that but the last time we discussed this, we believe that in a single biom (from a single preprocessing) it's not possible to get the same sequence with different casing, cc-ing: @ElDeveloper the main source of this information. Note that is possible that in a meta-analysis the same sequence can be found with different casing.

Comment (sjanssen2) [reply support_files/patches/180214_fix_upper_case.py]: Theoretically, I agree. But I think it wouldn't hurt if we compare the set size difference between current and updated. It there is no difference - as expected - good, but if there is we don't want to apply this to our life data! Good that we have the RC version where we can execute it first.

Comment (wasade) [reply support_files/patches/180214_fix_upper_case.py]: @rnaer, I think you mentioned you observed this the other day?

Comment (ElDeveloper) [reply support_files/patches/180214_fix_upper_case.py]: > ... we believe that in a single biom (from a single preprocessing) it's not possible to get the same sequence with different casing

hmmm I don't think we have any guarantee that this won't happen. I think we just assumed that deblur wouldn't do this. So it's not really a guarantee.

File: support_files/patches/180214_fix_upper_case.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+from qiita_db.sql_connection import TRN
+from biom.util import biom_open
+from skbio.io import read
+from os import rename, remove
+
+
+studies = Study.get_by_status('private').union(
+    Study.get_by_status('public')).union(Study.get_by_status('sandbox'))
+
+sft = Software.from_name_and_version('deblur', '1.0.4')
```
Comment (ElDeveloper) [inline support_files/patches/180214_fix_upper_case.py]: this variable is being overwritten.

File: support_files/patches/180214_fix_upper_case.py (hunk 1 above)
Comment (RNAer) [reply support_files/patches/180214_fix_upper_case.py]: I saw the same sequence in 2 observations like 'ACGt' and 'ACGT' in the latest AGP biom table - but I got it from Austin and don't know if this table is from single processing or merged.

File: support_files/patches/180214_deblur_v1.0.3_to_v1.0.4.sql [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            SET command_parameters = params
+            WHERE processing_job_id = pj_id;
+
+        UPDATE qiita.artifact
+            SET command_parameters = params
+            WHERE artifact_id = a_info.artifact_id;
+
+    END LOOP;
+    UPDATE qiita.command_parameter
+      SET check_biom_merge
```
Comment (josenavas) [inline support_files/patches/180214_deblur_v1.0.3_to_v1.0.4.sql]: I think you have to put which value are you setting it to?

File: support_files/patches/180214_fix_upper_case.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            tmp = fna + '.tmp'
+            with open(tmp, 'w') as out:
+                for seq in read(fna, format='fasta'):
+                    seq = str(seq)
+                    sequ = seq.upper()
+                    out.write('>%s\n%s\n' % (sequ, sequ))
+            rename(tmp, fna)
+            checksum = compute_checksum(fna)
+
+            TRN.add(sql, [checksum, ftps['preprocessed_fasta'][0]])
```
Comment (josenavas) [inline support_files/patches/180214_fix_upper_case.py]: Can you add the `TRN.execute()` at the end?

File: support_files/patches/180214_fix_upper_case.py (hunk 1 above)
Comment (antgonza) [reply support_files/patches/180214_fix_upper_case.py]: OK, I'll add some code for this.

File: support_files/patches/180214_deblur_v1.0.3_to_v1.0.4.sql (hunk 3 above)
Comment (antgonza) [reply support_files/patches/180214_deblur_v1.0.3_to_v1.0.4.sql]: I think you are right! 👅

File: support_files/patches/180214_fix_upper_case.py (hunk 4 above)
Comment (antgonza) [reply support_files/patches/180214_fix_upper_case.py]: yup

**Comment by antgonza** [comment]
I _think_ this is ready for another review.

File: support_files/patches/180214_fix_upper_case.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            tmp = fna + '.tmp'
+            with open(tmp, 'w') as out:
+                for seq in t.ids('observation'):
+                    out.write('>%s\n%s\n' % (seq, seq))
+            rename(tmp, fna)
+            checksum = compute_checksum(fna)
+
+            TRN.add(sql, [checksum, ftps['preprocessed_fasta'][0]])
+
+            TRN.execute()
```
Comment (sjanssen2) [inline support_files/patches/180214_fix_upper_case.py]: you only execute the transaction in cases where feature sequences need to be changed. Does that account for updating the deblur version number to 1.0.4 for the tables that need not to be changed?

Comment (antgonza) [reply support_files/patches/180214_fix_upper_case.py]: The change of version actually happens in the [SQL](https://github.com/qiita-spots/qp-deblur/pull/45/files#diff-513ae800d8385251d2557743d5d2e823R32). Here we are just fixing the upper/lower case.

**Comment by antgonza** [comment]
The SQL has been tested in qiitarc and now works as expected.

**Comment by sjanssen2** [comment]
@antgonza you have still some flake8 issues :-/

**Comment by antgonza** [comment]
Thanks! Forgot that this is python3 ...

**Pull request merged**